### PR TITLE
add serialization api based on ByteBuffer, #20324

### PR DESCRIPTION
**Pull request by patriknw**
- new trait ByteBufferSerializer with fromBinary and toBinary
  methods that takes ByteBuffer, this can be mixed in to
  existing serializer without breaking compatibility
- implement the ByteBufferSerializer in the ByteArraySerializer
- minor adjustment of the class manifest cache

### CODE REVIEW

**Comment by patriknw** [comment]
I'll measure any performance diff after lunch, but I expect it to be the same for byte arrays since essentially the same thing is done.

File: akka-remote/src/main/scala/akka/remote/MessageSerializer.scala [hunk 1]
```diff
@@ -53,27 +54,28 @@ private[akka] object MessageSerializer {
 
     // FIXME: This should be a FQCN instead
     headerBuilder.serializer = serializer.identifier.toString
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/MessageSerializer.scala]: @drewhk I'm not convinced that it's a good idea to use FQCN for the serializer identifier. I think that can become a problem for rolling upgrade scenarios where you want to rename serializer classes.

Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/MessageSerializer.scala]: OK, we can keep the numbering scheme, or alternatively, give names to the serializers.

**Comment by patriknw** [comment]
genjavadoc build error

File: akka-actor-tests/src/test/scala/akka/serialization/SerializeSpec.scala [hunk 2]
```diff
@@ -247,7 +248,26 @@ class SerializeSpec extends AkkaSpec(SerializationTests.serializeConf) {
...
       intercept[IllegalArgumentException] {
         byteSerializer.toBinary("pigdog")
-      }.getMessage should ===("ByteArraySerializer only serializes byte arrays, not [pigdog]")
+      }.getMessage should ===(s"${classOf[ByteArraySerializer].getName} only serializes byte arrays, not [java.lang.String]")
+    }
+
+    "support ByteBuffer serialization for byte arrays" in {
+      val byteSerializer = ser.serializerFor(classOf[Array[Byte]]).asInstanceOf[ByteBufferSerializer]
+
+      val byteBuffer = ByteBuffer.allocate(128).order(ByteOrder.LITTLE_ENDIAN)
```
Comment (drewhk) [inline akka-actor-tests/src/test/scala/akka/serialization/SerializeSpec.scala]: good

File: akka-actor-tests/src/test/scala/akka/serialization/SerializeSpec.scala [hunk 3]
```diff
@@ -247,7 +248,26 @@ class SerializeSpec extends AkkaSpec(SerializationTests.serializeConf) {
...
+      val str = "abcdef"
+      val payload = str.getBytes("UTF-8")
+      byteSerializer.toBinary(payload, byteBuffer)
+      byteBuffer.position() should ===(payload.length)
+      byteBuffer.flip()
+      byteBuffer.remaining() should ===(payload.length)
```
Comment (drewhk) [inline akka-actor-tests/src/test/scala/akka/serialization/SerializeSpec.scala]: this line is superfluous as it only tests flip() which I hope is tested already :)

File: akka-actor/src/main/scala/akka/serialization/Serialization.scala [hunk 4]
```diff
@@ -118,28 +121,60 @@ class Serialization(val system: ExtendedActorSystem) extends Extension {
...
+   */
+  def deserializeByteBuffer(buf: ByteBuffer, serializerId: Int, manifest: String): AnyRef = {
+    val serializer = try serializerByIdentity(serializerId) catch {
+      case _: NoSuchElementException ⇒ throw new NotSerializableException(
+        s"Cannot find serializer with id [$serializerId]. The most probable reason is that the configuration entry " +
+          "akka.actor.serializers is not in synch between the two systems.")
+    }
+    serializer match {
+      case ser: ByteBufferSerializer =>
+        ser.fromBinary(buf, manifest)
```
Comment (drewhk) [inline akka-actor/src/main/scala/akka/serialization/Serialization.scala]: nice

**Comment by drewhk** [comment]
LGTM

File: akka-actor-tests/src/test/scala/akka/serialization/SerializeSpec.scala (hunk 3 above)
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/serialization/SerializeSpec.scala]: yeah, I was a bit uncertain myself about those lengths, I'll remove

**Comment by patriknw** [comment]
@drewhk I have added a few more things related to serialization here, see:

* e74e1da 2016-05-27 optimize currentTransportInformation
* aa2c4fe 2016-05-27 handle serialization errors, #20324
* a0f8f19 2016-05-27 change header to use serializer id instead of FQCN

**Comment by patriknw** [comment]
@ktoso note that with a0f8f19 we only use serializer id in the header, it will not be in the compression table

**Pull request merged**